dataplane/digest/value.go

```diff
@@ -0,0 +1,371 @@
+package digest
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/neblic/platform/controlplane/control"
+	"github.com/neblic/platform/dataplane/protos"
+	"github.com/neblic/platform/internal/pkg/data"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type ValueType int8
+
+const (
+	UnknownValueType = ValueType(iota)
+	NumberValueType
+	StringValueType
+	BooleanValueType
+	ObjValueType
+	ArrayValueType
+	NilValueType
+)
+
+type Value struct {
+	opts ValueOptions
+
+	fieldsProcessed int
+	digest          *protos.ObjValue
+}
+
+func (v *Value) isDigest() {}
+
+func NewValue(opts ValueOptions) *Value {
+	return &Value{
+		opts: opts,
+
+		digest: protos.NewObjValue(),
+	}
+}
+
+func (v *Value) updateMinStat(state *protos.MinStat, value float64) {
+	state.Value = math.Min(state.Value, value)
+}
+
+func (v *Value) updateAvgStat(state *protos.AvgStat, value float64) {
+	state.Sum += value
+	state.Count++
+}
+
+func (v *Value) updateMaxStat(state *protos.MaxStat, value float64) {
+	state.Value = math.Max(state.Value, value)
+}
+
+func (v *Value) updateBoolean(state *protos.BooleanValue, boolean *bool) (*protos.BooleanValue, error) {
+	if err := v.incrFieldsProcessed(); err != nil {
+		return nil, err
+	}
+
+	if boolean != nil {
+		if *boolean {
+			state.TotalTrue++
+		} else {
+			state.DefaultCount++
+			state.TotalFalse++
+		}
+	} else {
+		state.NullCount++
+	}
+	state.TotalCount++
+
+	return state, nil
+}
+
+func (v *Value) updateNum(state *protos.NumberValue, number *float64) (*protos.NumberValue, error) {
+	if err := v.incrFieldsProcessed(); err != nil {
+		return nil, err
+	}
+
+	if number != nil {
+		if *number == 0.0 {
+			state.DefaultCount++
+		}
+		v.updateMinStat(state.Min, *number)
+		v.updateAvgStat(state.Avg, *number)
+		v.updateMaxStat(state.Max, *number)
+	} else {
+		state.NullCount++
+	}
+	state.TotalCount++
+
+	return state, nil
+
+}
+
+func (v *Value) updateString(state *protos.StringValue, str *string) (*protos.StringValue, error) {
+	if err := v.incrFieldsProcessed(); err != nil {
+		return nil, err
+	}
+
+	if str != nil {
+		if *str == "" {
+			state.DefaultCount++
+		}
+		v.updateMinStat(state.Length.Min, float64(len(*str)))
+		v.updateAvgStat(state.Length.Avg, float64(len(*str)))
+		v.updateMaxStat(state.Length.Max, float64(len(*str)))
+
+	} else {
+		state.NullCount++
+	}
+	state.TotalCount++
+
+	return state, nil
+}
+
+func (v *Value) updateArray(state *protos.ArrayValue, array []interface{}) (*protos.ArrayValue, error) {
+	if err := v.incrFieldsProcessed(); err != nil {
+		return nil, err
+	}
+
+	if array != nil {
+		if len(array) == 0 {
+			state.DefaultCount++
+		}
+
+		// Just one digest is kept for the full array, update it with all the array entries
+		for _, value := range array {
+			_, err := v.updateValue(state.Values, value)
```

Comment:
We need to implement a similar logic in the structure digest. Otherwise, it will be really hard to match the structure to the value digests.

NOTE: I moved the comment from its previous location since I put it in the wrong place

Reply:
to be done in a follow-up PR